{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "19a484da-4f81-494a-b279-6778904bdc4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting dataframes.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile dataframes.py\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "class CoronaDataframe:\n",
    "    '''Class that handles manipulation and generation of dataframes\n",
    "    needed for plotting'''\n",
    "    \n",
    "    def __init__(self, query: str, ref: str, mutations: str) -> None:\n",
    "        \n",
    "        # Only include region between 256:29674 (same as pangolin)\n",
    "        self.query = SeqIO.read(query, 'fasta').seq[256:29674]\n",
    "        self.ref = SeqIO.read(ref, 'fasta').seq[256:29674]\n",
    "        self.mutations = mutations\n",
    "    \n",
    "    \n",
    "    def make_alignment_df(self) -> pd.DataFrame:\n",
    "        '''Turns the sequence of the aligned fasta file into a df used to plot \n",
    "        the missing (N), deletions and mutations as a genome map'''\n",
    "    \n",
    "        reference = [x for x in self.ref]\n",
    "        query = [x for x in self.query]\n",
    "        converted_values = self._convert_alignment_value()\n",
    "\n",
    "        alignment_df = pd.DataFrame({'nt': query, 'id': 'query', \n",
    "                                 'pos': range(256, len(query) + 256), \n",
    "                                 'value': converted_values, \n",
    "                                 'ref': reference})\n",
    "        return alignment_df\n",
    "    \n",
    "    \n",
    "    def make_mutations_df(self) -> pd.DataFrame:\n",
    "        '''Transforms the df of mutations to a suitable format'''\n",
    "        \n",
    "        mutations_df = pd.read_csv(self.mutations)\n",
    "        mutations_df['unique'] = [1 if x == 1 else 0 for x in mutations_df.number_mutations]\n",
    "        mutations_df['kind'] = [2 if '-' in str(x) else y for x,y in zip(mutations_df.mutation, \n",
    "                                                                         mutations_df.unique)]\n",
    "        mutations_df['position'] = mutations_df['mutation'].str.extract('(\\d+)').astype(int)\n",
    "\n",
    "        return mutations_df\n",
    "    \n",
    "    \n",
    "    def make_query_df(self) -> pd.DataFrame:\n",
    "        '''Turns a list of mutations for the query into a df in the right format'''\n",
    "        \n",
    "        mutations_list = self._extract_snp()\n",
    "        query_df = pd.DataFrame({'pango': 'QUERY', 'mutation': mutations_list})\n",
    "        query_df['position'] = query_df['mutation'].str.extract('(\\d+)').astype(int)\n",
    "    \n",
    "        return query_df\n",
    "    \n",
    "    \n",
    "    def make_mutations_inspection_df(self) -> pd.DataFrame:\n",
    "        '''Returns a dataframe with all mutations and lineages \n",
    "        for every given position in the query mutation set'''\n",
    "        mutations = self.make_mutations_df()\n",
    "        query_df = self.make_query_df()\n",
    "        mutations = mutations[mutations['position'].isin(query_df.position)]\n",
    "        query_df = query_df[query_df['position'].isin(mutations.position)]\n",
    "\n",
    "\n",
    "        mutations_list = mutations.groupby('position')[['mutation', 'pango']].aggregate(list).reset_index()\n",
    "        mutations_list['mutation_list'] = [dict(zip(x, y)) for x,y in zip(mutations_list.pango, mutations_list.mutation)]\n",
    "        mutations_list.drop(columns=['pango', 'mutation'], inplace=True)\n",
    "\n",
    "        merged = query_df.merge(mutations_list)\n",
    "        merged['pango_with_mutation'] = merged['mutation_list'].apply(len)\n",
    "        merged['color'] = [1 if x.endswith('N') else 2 if x.endswith('-') else 3 for x in merged.mutation]\n",
    "\n",
    "        return merged\n",
    "    \n",
    "    \n",
    "    def _convert_alignment_value(self) -> list:\n",
    "        '''Converts alignments to numbers depending on it is N, \n",
    "        deletion, substitution or the same'''\n",
    "        values = []\n",
    "        for q,r in zip(self.query, self.ref):\n",
    "            if q != r:\n",
    "                if q == 'N':\n",
    "                    values.append(-1)\n",
    "                elif q == '-':\n",
    "                    values.append(-2)\n",
    "                else:\n",
    "                    values.append(1)\n",
    "            else:\n",
    "                values.append(0)  \n",
    "        return values\n",
    "    \n",
    "    #### MÅSTE ADDERA +1 till index senare!!!!!!!!!!! #######\n",
    "    def _extract_snp(self) -> list:\n",
    "        '''Helper function to extract information about mutation'''\n",
    "\n",
    "        return [f'{a}{index}{b}' for index,(a,b) in \n",
    "                enumerate(zip(self.ref, self.query), 256) if a != b]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4349b07e-0f42-42c8-92d4-65026da93233",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting plotting.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile plotting.py\n",
    "import altair as alt\n",
    "import pandas as pd\n",
    "alt.data_transformers.disable_max_rows()\n",
    "\n",
    "from .dataframes import CoronaDataframe\n",
    "\n",
    "class CoronaPlot:\n",
    "    '''Class for plotting corona data. Uses Altair.'''\n",
    "    \n",
    "    def __init__(self, query: str, ref: str, mutations: str) -> None:\n",
    "        self.dataframes = CoronaDataframe(query, ref, mutations)\n",
    "        \n",
    "    def plot_alignment_map(self):\n",
    "        ''' Returns chart and view which plots the alignment map of the query sequence'''\n",
    "        ####### Chache with json file????? ######## \n",
    "        df = self.dataframes.make_alignment_df()\n",
    "        interval = alt.selection_interval(encodings=['x'])\n",
    "\n",
    "        base = alt.Chart(df, title=\"Coverage and mutations\").mark_rect().encode(\n",
    "            x=alt.X('pos:N', axis=alt.Axis(labels=False, ticks=False)),\n",
    "            y='id:N',\n",
    "            color=alt.Color('max(value):N', legend=None, scale=alt.Scale(domain=[0,-1, -2, 1], range=['white', 'red', 'black', 'green'])),\n",
    "            tooltip=[alt.Tooltip('nt:N', title='nt'), alt.Tooltip('pos:N', title='Position'), alt.Tooltip('ref:N', title='Reference')])\n",
    "\n",
    "        chart = base.encode(x=alt.X('pos:N', scale=alt.Scale(domain=interval.ref()), axis=alt.Axis(labels=False, ticks=False))\n",
    "            ).properties(width=1200, height=300\n",
    "            )\n",
    "\n",
    "        view = base.add_selection(interval\n",
    "        ).properties(width=1200, height=100)\n",
    "\n",
    "\n",
    "        return chart, view\n",
    "        \n",
    "        \n",
    "    def plot_mutation_heatmap(self):\n",
    "        '''Plots a heatmap over all mutations in the mutation csv file'''\n",
    "        mutations = self.dataframes.make_mutations_df()\n",
    "\n",
    "        unique_mut = list(mutations.mutation.unique())\n",
    "        order_level = sorted(unique_mut, key=lambda x: int(\"\".join([i for i in x if i.isdigit()])))\n",
    "\n",
    "        fig = alt.Chart(mutations, title='Mutations. Dark blue == deletions. Light blue == unique mutations for that lineage').mark_rect(stroke='black').encode(\n",
    "            x=alt.X('mutation', axis=alt.Axis(labels=False, ticks=False), sort=order_level),\n",
    "            y='pango',\n",
    "            color=alt.Color('kind', legend=None),\n",
    "            tooltip=[\n",
    "                alt.Tooltip('mutation', title='Mutation'),\n",
    "                alt.Tooltip('pango', title='Pango')]).properties(width=1200)\n",
    "\n",
    "        return fig\n",
    "    \n",
    "    \n",
    "    \n",
    "    def plot_compare(self, compare_to: str):\n",
    "        '''Compares query mutatations against mutations for a given corona lineage.'''\n",
    "        mutations_db = self.dataframes.make_mutations_df()\n",
    "        mutations_db = mutations_db[['pango', 'mutation', 'kind', 'position']]\n",
    "        lineage_df = mutations_db[mutations_db['pango'] == compare_to]\n",
    "\n",
    "        query_df = self.dataframes.make_query_df()\n",
    "        # Vad är bäst, jämföra mot bara mutationer i den man vill jämföra med eller ta alla mutationer som finns i query???\n",
    "        # kommentera bort raderna nedan för att testa\n",
    "        #query_df = query_df[query_df['position'].isin(mutations_db.position.to_list())]\n",
    "        query_df = query_df[query_df['position'].isin(lineage_df.position.to_list())]\n",
    "\n",
    "        query_df['kind'] = [3 if x in lineage_df.mutation.to_list() else 4 for x in query_df.mutation]\n",
    "\n",
    "        concated = pd.concat([lineage_df, query_df])\n",
    "        position_level = sorted(concated.position.unique())\n",
    "\n",
    "        fig = alt.Chart(concated, title={'text': [f'Compared against {compare_to}'], \n",
    "                                         'subtitle': ['Green == same, pink == unique, blue == deletion, red == differs from comparison'], \n",
    "                                         'color': 'green'}).mark_rect(stroke='black').encode(\n",
    "        x=alt.X('position:N', axis=alt.Axis(), sort=position_level),\n",
    "        y='pango',\n",
    "        color=alt.Color('kind', legend=None, scale=alt.Scale(domain=[0, 1, 2, 3, 4], range=['#e7fc98 ', '#f64fe7', '#98fcf3', '#e7fc98', '#fc9898'])),\n",
    "        tooltip=[\n",
    "            alt.Tooltip('mutation', title='Mutation'),\n",
    "            alt.Tooltip('pango', title='Pango')]).properties(width=1000, height=50)\n",
    "\n",
    "        return fig\n",
    "\n",
    "\n",
    "    def plot_mutations_inspection(self):\n",
    "        '''Plots how common each mutation is for every mutation in the query sequence'''\n",
    "        df = self.dataframes.make_mutations_inspection_df()\n",
    "\n",
    "        fig = alt.Chart(df, title={'text': ['Mutations in'], \n",
    "                                         'subtitle': ['Green == regular, blue == deletion, red == N', \n",
    "                                                      'Size shows how common mutation is'], \n",
    "                                         'color': 'green'}).mark_point(stroke='black').encode(\n",
    "            x=alt.X('position:N', axis=alt.Axis()),\n",
    "            y='pango',\n",
    "            fill=alt.Fill('color:N', legend=None, scale=alt.Scale(domain=[1, 2, 3], range=['red', 'blue', 'green'])),\n",
    "            size=alt.Size('pango_with_mutation', legend=None, scale=alt.Scale(range=[100, 1500])),\n",
    "            tooltip=[\n",
    "                alt.Tooltip('mutation', title='Mutation'),\n",
    "                alt.Tooltip('mutation_list', title='Pangos with mutation')]).properties(width=1200, height=150)\n",
    "\n",
    "        return fig\n",
    "        \n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6567125c-df43-4705-a8d4-63a72a661aeb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting nextclade.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile nextclade.py\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "\n",
    "class Nextclade:\n",
    "    '''Runs nextclade and stores the information of the run'''\n",
    "    \n",
    "    def __init__(self) -> None:\n",
    "        self.OUTDIR = 'nextclade_runs/'\n",
    "        self.REF_DIR = 'reference/'\n",
    "        self.INPUT = 'input.fasta'\n",
    "        self.META_TSV = self.OUTDIR + 'meta.tsv'\n",
    "        self.QUERY = self.OUTDIR + 'input.aligned.fasta'\n",
    "        self.REF = self.REF_DIR + 'reference.fasta'\n",
    "        \n",
    "    def run_nextclade(self) -> None: \n",
    "        command = ['nextclade', '--in-order', '--input-fasta', self.INPUT, '--input-dataset', self.REF_DIR,\n",
    "                   '--output-dir', self.OUTDIR, '--output-tsv', self.META_TSV]\n",
    "\n",
    "        subprocess.call(command)\n",
    "        \n",
    "    def make_meta_df(self) -> None:\n",
    "        return pd.read_csv(self.META_TSV, sep='\\t')\n",
    "        \n",
    "        \n",
    "        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
